server/src/main/java/io/crate/planner/optimizer/rule/MoveFilterBeneathJoin.java

```diff
@@ -32,27 +51,64 @@
 import io.crate.planner.optimizer.matcher.Captures;
 import io.crate.planner.optimizer.matcher.Pattern;
 
-import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
-import static io.crate.planner.optimizer.matcher.Patterns.source;
-import static io.crate.planner.optimizer.rule.FilterOnJoinsUtil.moveQueryBelowJoin;
-
-import java.util.function.Function;
-
+/**
+ * If we can determine that a filter on an OUTER JOIN turns all NULL rows that the join could generate into a NO-MATCH
+ * we can push down the filter and turn the OUTER JOIN into an inner join.
+ *
+ * <p>
+ * So this tries to transform
+ * </p>
+ *
+ * <pre>
+ *     Filter (lhs.x = 1 AND rhs.x = 2)
+ *       |
+ *     Outer-Join
+ *       /  \
+ *     LHS  RHS
+ * </pre>
+ *
+ * into
+ *
+ * <pre>
+ *     Filter
+ *       |
+ *     Inner-Join
+ *       /      \
+ *   Filter      Filter
+ * (lhs.x = 1)    (rhs.x = 2)
+ *     |              |
+ *    LHS            RHS
+ * </pre>
+ *
+ * A case where this is *NOT* safe is for example:
+ *
+ * <pre>
+ * SELECT * FROM t1
+ *     LEFT JOIN t2 ON t1.t2_id = t2.id
+ * WHERE
+ *     coalesce(t2.x, 20) > 10   # becomes TRUE for null rows
+ * </pre>
+ *
+ *
+ * A case where the FILTER turns all null rows into no-matches:
+ *
+ * <pre>
+ * SELECT * FROM t1
+ *      LEFT JOIN t2 ON t1.t2_id = t2.id
+ * WHERE
+ *      t2.x = 10           # null = 10 -> null -> no match
+ * </pre>
+ */
 public final class MoveFilterBeneathJoin implements Rule<Filter> {
 
-    private final Capture<AbstractJoinPlan> joinCapture;
+    private final Capture<JoinPlan> nlCapture;
     private final Pattern<Filter> pattern;
 
     public MoveFilterBeneathJoin() {
-        this.joinCapture = new Capture<>();
+        this.nlCapture = new Capture<>();
         this.pattern = typeOf(Filter.class)
-            .with(source(),
-                  typeOf(AbstractJoinPlan.class)
-                      .capturedAs(joinCapture)
-                      // Can't apply this on OUTER JOINs as outer join actively produce new null rows
-                      // We need to run the filter on top of these null rows to produce the correct results
-                      .with(join -> !join.joinType().isOuter())
-            );
+                .with(source(), typeOf(JoinPlan.class).capturedAs(nlCapture)
```

Comment:
If we change this to `JoinPlan`, `NestedLoopJoins` and "HashJoin` will not match anymore.